File: ios/ArgumentAdapter.m
//
//  ArgumentAdapter.cpp
//  rongcloud_rtc_wrapper_plugin
//
//  Created by 潘铭达 on 2021/6/11.
//

#import "ArgumentAdapter.h"

RCRTCIWMediaType toMediaType(int type) {
    return (RCRTCIWMediaType) type;
}

RCRTCIWCamera toCamera(int camera) {
    return (RCRTCIWCamera) (camera - 1);
}

RCRTCIWCameraCaptureOrientation toCameraCaptureOrientation(int orientation) {
    return (RCRTCIWCameraCaptureOrientation) (orientation + 1);
}

RCRTCIWLiveMixLayoutMode toLiveMixLayoutMode(int mode) {
    return (RCRTCIWLiveMixLayoutMode) (mode + 1);
}

RCRTCIWLiveMixRenderMode toLiveMixRenderMode(int mode) {
    return (RCRTCIWLiveMixRenderMode) (mode + 1);
}

RCRTCIWVideoResolution toVideoResolution(int resolution) {
    return (RCRTCIWVideoResolution) resolution;
}

RCRTCIWVideoFps toVideoFps(int fps) {
    return (RCRTCIWVideoFps) fps;
}

RCRTCIWAudioMixingMode toAudioMixingMode(int mode) {
    return (RCRTCIWAudioMixingMode) mode;
}

RCRTCIWAudioCodecType toAudioCodecType(int type) {
    switch (type) {
        case 1:
            return RCRTCIWAudioCodecTypeOPUS;
        default:
            return RCRTCIWAudioCodecTypePCMU;
    }
}

RCRTCIWRole toRole(int role) {
    return (RCRTCIWRole) role;
}

RCRTCIWAudioQuality toAudioQuality(int quality) {
    switch (quality) {
        case 1:
            return RCRTCIWAudioQualityMusic;
        case 2:
            return RCRTCIWAudioQualityMusicHigh;
        default:
            return RCRTCIWAudioQualitySpeech;
    }
}

RCRTCIWAudioScenario toAudioScenario(int scenario) {
    return (RCRTCIWAudioScenario) scenario;
}

RCRTCIWStreamType toStreamType(int type) {
    return (RCRTCIWStreamType) (type - 1);
}

int getInt(NSDictionary *arguments, NSString *key, int defaultValue) {
    if (![arguments[key] isEqual:[NSNull null]]) {
        return [arguments[key] intValue];
    }
    return defaultValue;
}

bool getBool(NSDictionary *arguments, NSString *key, bool defaultValue) {
    if (![arguments[key] isEqual:[NSNull null]]) {
        return [arguments[key] boolValue];
    }
    return defaultValue;
}

RCRTCIWAudioSetup *toAudioSetup(NSDictionary *arguments) {
    int codec = getInt(arguments, @"codec", 1); // default opus
    RCRTCIWAudioSetup *setup = [[RCRTCIWAudioSetup alloc] init];
    setup.type = toAudioCodecType(codec);
    return setup;
}

RCRTCIWVideoSetup *toVideoSetup(NSDictionary *arguments) {
    bool enableTinyStream = getBool(arguments, @"enableTinyStream", true);
    RCRTCIWVideoSetup *setup = [[RCRTCIWVideoSetup alloc] init];
    setup.enableTinyStream = enableTinyStream;
    return setup;
}

RCRTCIWEngineSetup *toEngineSetup(NSDictionary *arguments) {
    bool reconnectable = getBool(arguments, @"reconnectable", true);
    int statsReportInterval = getInt(arguments, @"statsReportInterval", 1000);
    id mediaUrl = arguments[@"mediaUrl"];
    id audioSetup = arguments[@"audioSetup"];
    id videoSetup = arguments[@"videoSetup"];
    RCRTCIWEngineSetup *setup = [[RCRTCIWEngineSetup alloc] init];
    setup.reconnectable = reconnectable;
    setup.statsReportInterval = statsReportInterval;
    if (![mediaUrl isEqual:[NSNull null]]) {
        setup.mediaUrl = mediaUrl;
    }
    if (![audioSetup isEqual:[NSNull null]]) {
        setup.audioSetup = toAudioSetup(audioSetup);
    }
    if (![videoSetup isEqual:[NSNull null]]) {
        setup.videoSetup = toVideoSetup(videoSetup);
    }
    return setup;
}

RCRTCIWRoomSetup *toRoomSetup(NSDictionary *arguments) {
    int type = getInt(arguments, @"type", 2); // default audio_video
    int role = getInt(arguments, @"role", 0); // default meeting_member
    RCRTCIWRoomSetup *setup = [[RCRTCIWRoomSetup alloc] init];
    setup.type = toMediaType(type);
    setup.role = toRole(role);
    return setup;
}

RCRTCIWAudioConfig *toAudioConfig(NSDictionary *arguments) {
    int quality = getInt(arguments, @"quality", 0); // speech
    int scenario = getInt(arguments, @"scenario", 0); // normal
    RCRTCIWAudioConfig *config = [[RCRTCIWAudioConfig alloc] init];
    config.quality = toAudioQuality(quality);
    config.scenario = toAudioScenario(scenario);
    return config;
}

RCRTCIWVideoConfig *toVideoConfig(NSDictionary *arguments) {
    int minBitrate = getInt(arguments, @"minBitrate", 250);
    int maxBitrate = getInt(arguments, @"maxBitrate", 2200);
    int fps = getInt(arguments, @"fps", 2); // default 25fps
    int resolution = getInt(arguments, @"resolution", 15); // default 720p
    RCRTCIWVideoConfig *config = [[RCRTCIWVideoConfig alloc] init];
    config.minBitrate = minBitrate;
    config.maxBitrate = maxBitrate;
    config.fps = toVideoFps(fps);
    config.resolution = toVideoResolution(resolution);
    return config;
}

RCRTCIWCustomLayout *toLiveMixCustomLayout(NSDictionary *arguments) {
    NSString *uid = arguments[@"id"];
    int x = getInt(arguments, @"x", 0);
    int y = getInt(arguments, @"y", 0);
    int width = getInt(arguments, @"width", 480);
    int height = getInt(arguments, @"height", 640);
    RCRTCIWStreamType type = toStreamType([arguments[@"type"] intValue]);
    NSString *tag = arguments[@"tag"];
    RCRTCIWCustomLayout *layout = [[RCRTCIWCustomLayout alloc] initWithStreamType:type userId:uid tag:tag];
    layout.x = x;
    layout.y = y;
    layout.width = width;
    layout.height = height;
    return layout;
}

NSArray<RCRTCIWCustomLayout *> *toLiveMixCustomLayouts(NSArray<NSDictionary *> *arguments) {
    NSMutableArray *layouts = [NSMutableArray array];
    for (NSDictionary *argument in arguments) {
        [layouts addObject:toLiveMixCustomLayout(argument)];
    }
    return layouts;
}

NSDictionary *fromNetworkStats(RCRTCIWNetworkStats *stats) {
    NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];
    [dictionary setObject:@((int) stats.type) forKey:@"type"];
    [dictionary setObject:stats.ip forKey:@"ip"];
    [dictionary setObject:@(stats.sendBitrate) forKey:@"sendBitrate"];
    [dictionary setObject:@(stats.receiveBitrate) forKey:@"receiveBitrate"];
    [dictionary setObject:@(stats.rtt) forKey:@"rtt"];
    return dictionary;
}

int fromAudioCodecType(RCRTCIWAudioCodecType type) {
    switch (type) {
        case RCRTCIWAudioCodecTypePCMU:
            return 0;
        default:
            return 1;
    }
}

NSDictionary *fromLocalAudioStats(RCRTCIWLocalAudioStats *stats) {
    NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];
    [dictionary setObject:@(fromAudioCodecType(stats.codec)) forKey:@"codec"];
    [dictionary setObject:@(stats.bitrate) forKey:@"bitrate"];
    [dictionary setObject:@(stats.volume) forKey:@"volume"];
    [dictionary setObject:@(stats.packageLostRate) forKey:@"packageLostRate"];
    return dictionary;
}

NSDictionary *fromLocalVideoStats(RCRTCIWLocalVideoStats *stats) {
    NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];
    [dictionary setObject:@(stats.tiny) forKey:@"tiny"];
    [dictionary setObject:@((int) stats.codec) forKey:@"codec"];
    [dictionary setObject:@(stats.bitrate) forKey:@"bitrate"];
    [dictionary setObject:@(stats.fps) forKey:@"fps"];
    [dictionary setObject:@(stats.width) forKey:@"width"];
    [dictionary setObject:@(stats.height) forKey:@"height"];
    [dictionary setObject:@(stats.packageLostRate) forKey:@"packageLostRate"];
    return dictionary;
}

NSDictionary *fromRemoteAudioStats(RCRTCIWRemoteAudioStats *stats) {
    NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];
    [dictionary setObject:@(fromAudioCodecType(stats.codec)) forKey:@"codec"];
    [dictionary setObject:@(stats.bitrate) forKey:@"bitrate"];
    [dictionary setObject:@(stats.volume) forKey:@"volume"];
    [dictionary setObject:@(stats.packageLostRate) forKey:@"packageLostRate"];
    [dictionary setObject:@(stats.rtt) forKey:@"rtt"];
    return dictionary;
}

NSDictionary *fromRemoteVideoStats(RCRTCIWRemoteVideoStats *stats) {
    NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];
    [dictionary setObject:@((int) stats.codec) forKey:@"codec"];
    [dictionary setObject:@(stats.bitrate) forKey:@"bitrate"];
    [dictionary setObject:@(stats.fps) forKey:@"fps"];
    [dictionary setObject:@(stats.width) forKey:@"width"];
    [dictionary setObject:@(stats.height) forKey:@"height"];
    [dictionary setObject:@(stats.packageLostRate) forKey:@"packageLostRate"];
    [dictionary setObject:@(stats.rtt) forKey:@"rtt"];
    return dictionary;
}

File: android/src/main/java/cn/rongcloud/rtc/rn/ArgumentAdapter.java
package cn.rongcloud.rtc.rn;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;

import java.util.ArrayList;
import java.util.List;

import cn.rongcloud.rtc.wrapper.config.RCRTCIWAudioConfig;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWStreamType;
import cn.rongcloud.rtc.wrapper.module.RCRTCIWCustomLayout;
import cn.rongcloud.rtc.wrapper.config.RCRTCIWVideoConfig;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWAudioCodecType;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWAudioMixingMode;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWAudioQuality;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWAudioScenario;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWCamera;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWCameraCaptureOrientation;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWLiveMixLayoutMode;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWLiveMixRenderMode;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWLocalAudioStats;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWLocalVideoStats;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWMediaType;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWNetworkStats;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWRemoteAudioStats;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWRemoteVideoStats;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWRole;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWVideoFps;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWVideoResolution;
import cn.rongcloud.rtc.wrapper.setup.RCRTCIWAudioSetup;
import cn.rongcloud.rtc.wrapper.setup.RCRTCIWEngineSetup;
import cn.rongcloud.rtc.wrapper.setup.RCRTCIWRoomSetup;
import cn.rongcloud.rtc.wrapper.setup.RCRTCIWVideoSetup;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWNetworkProbeStats;

final class ArgumentAdapter {
  private ArgumentAdapter() {
  }

  static RCRTCIWMediaType toMediaType(@NonNull Integer type) {
    return RCRTCIWMediaType.values()[type];
  }

  static RCRTCIWCamera toCamera(@NonNull Integer camera) {
    return RCRTCIWCamera.values()[camera];
  }

  static RCRTCIWCameraCaptureOrientation toCameraCaptureOrientation(@NonNull Integer orientation) {
    return RCRTCIWCameraCaptureOrientation.values()[orientation];
  }

  static RCRTCIWLiveMixLayoutMode toLiveMixLayoutMode(@NonNull Integer mode) {
    return RCRTCIWLiveMixLayoutMode.values()[mode];
  }

  static RCRTCIWLiveMixRenderMode toLiveMixRenderMode(@NonNull Integer mode) {
    return RCRTCIWLiveMixRenderMode.values()[mode];
  }

  static RCRTCIWVideoResolution toVideoResolution(@NonNull Integer resolution) {
    return RCRTCIWVideoResolution.values()[resolution];
  }

  static RCRTCIWVideoFps toVideoFps(@NonNull Integer fps) {
    return RCRTCIWVideoFps.values()[fps];
  }

  static RCRTCIWAudioMixingMode toAudioMixingMode(@NonNull Integer mode) {
    return RCRTCIWAudioMixingMode.values()[mode];
  }

  static RCRTCIWAudioCodecType toAudioCodecType(@NonNull Integer type) {
    return RCRTCIWAudioCodecType.values()[type];
  }

  static RCRTCIWRole toRole(@NonNull Integer role) {
    return RCRTCIWRole.values()[role];
  }

  static RCRTCIWAudioQuality toAudioQuality(@NonNull Integer quality) {
    return RCRTCIWAudioQuality.values()[quality];
  }

  static RCRTCIWAudioScenario toAudioScenario(@NonNull Integer scenario) {
    return RCRTCIWAudioScenario.values()[scenario];
  }

  private static int getInt(@NonNull ReadableMap map, String key, int defaultValue) {
    if (map.hasKey(key)) {
      return map.getInt(key);
    }
    return defaultValue;
  }

  private static boolean getBoolean(@NonNull ReadableMap map, String key, boolean defaultValue) {
    if (map.hasKey(key)) {
      return map.getBoolean(key);
    }
    return defaultValue;
  }

  static RCRTCIWAudioSetup toAudioSetup(@NonNull ReadableMap setup) {
    int codec = getInt(setup, "codec", 1); // default opus
    int audioSource = getInt(setup, "audioSource", 7); // default VOICE_COMMUNICATION
    int audioSampleRate = getInt(setup, "audioSampleRate", 16000);
    boolean enableMicrophone = getBoolean(setup, "enableMicrophone", true);
    boolean enableStereo = getBoolean(setup, "enableStereo", true);
    return RCRTCIWAudioSetup.Builder.create()
      .withAudioCodecType(toAudioCodecType(codec))
      .withAudioSource(audioSource)
      .withAudioSampleRate(audioSampleRate)
      .withEnableMicrophone(enableMicrophone)
      .withEnableStereo(enableStereo)
      .build();
  }

  static RCRTCIWVideoSetup toVideoSetup(@NonNull ReadableMap setup) {
    boolean enableHardwareDecoder = getBoolean(setup, "enableHardwareDecoder", true);
    boolean enableHardwareEncoder = getBoolean(setup, "enableHardwareEncoder", true);
    boolean enableHardwareEncoderHighProfile = getBoolean(setup, "enableHardwareEncoderHighProfile", false);
    int hardwareEncoderFrameRate = getInt(setup, "hardwareEncoderFrameRate", 30);
    boolean enableEncoderTexture = getBoolean(setup, "enableEncoderTexture", true);
    boolean enableTinyStream = getBoolean(setup, "enableTinyStream", true);
    return RCRTCIWVideoSetup.Builder.create()
      .withEnableHardwareDecoder(enableHardwareDecoder)
      .withEnableHardwareEncoder(enableHardwareEncoder)
      .withEnableHardwareEncoderHighProfile(enableHardwareEncoderHighProfile)
      .withHardwareEncoderFrameRate(hardwareEncoderFrameRate)
      .withEnableEncoderTexture(enableEncoderTexture)
      .withEnableTinyStream(enableTinyStream)
      .build();
  }

  static RCRTCIWEngineSetup toEngineSetup(@NonNull ReadableMap setup) {
    boolean reconnectable = getBoolean(setup, "reconnectable", true);
    int statsReportInterval = getInt(setup, "statsReportInterval", 1000);
    RCRTCIWEngineSetup.Builder builder = RCRTCIWEngineSetup.Builder.create()
      .withReconnectable(reconnectable)
      .withStatsReportInterval(statsReportInterval);
    if (setup.hasKey("mediaUrl")) {
      builder.withMediaUrl(setup.getString("mediaUrl"));
    }
    if (setup.hasKey("audioSetup")) {
      ReadableMap audioSetup = setup.getMap("audioSetup");
      if (audioSetup != null) {
        builder.withAudioSetup(toAudioSetup(audioSetup));
      }
    }
    if (setup.hasKey("videoSetup")) {
      ReadableMap videoSetup = setup.getMap("videoSetup");
      if (videoSetup != null) {
        builder.withVideoSetup(toVideoSetup(videoSetup));
      }
    }
    return builder.build();
  }

  static RCRTCIWRoomSetup toRoomSetup(@NonNull ReadableMap setup) {
    int type = getInt(setup, "type", 2); // default audio_video
    int role = getInt(setup, "role", 0); // default meeting_member
    RCRTCIWRoomSetup.Builder builder = RCRTCIWRoomSetup.Builder.create()
      .withMediaType(toMediaType(type))
      .withRole(toRole(role));
    return builder.build();
  }

  static RCRTCIWAudioConfig toAudioConfig(@NonNull ReadableMap config) {
    int quality = getInt(config, "quality", 0); // default speech
    int scenario = getInt(config, "scenario", 0); // default normal
    return RCRTCIWAudioConfig.create()
      .setQuality(toAudioQuality(quality))
      .setScenario(toAudioScenario(scenario));
  }

  static RCRTCIWVideoConfig toVideoConfig(@NonNull ReadableMap config) {
    int minBitrate = getInt(config, "minBitrate", 250);
    int maxBitrate = getInt(config, "maxBitrate", 2200);
    int fps = getInt(config, "fps", 2); // default 25fps
    int resolution = getInt(config, "resolution", 15); // default 720p
    return RCRTCIWVideoConfig.create()
      .setMinBitrate(minBitrate)
      .setMaxBitrate(maxBitrate)
      .setFps(toVideoFps(fps))
      .setResolution(toVideoResolution(resolution));
  }

  static List<RCRTCIWCustomLayout> toLiveMixCustomLayouts(@NonNull ReadableArray layouts) {
    List<RCRTCIWCustomLayout> lists = new ArrayList<>(layouts.size());
    for (int i = 0; i < layouts.size(); i++) {
      lists.add(toLiveMixCustomLayout(layouts.getMap(i)));
    }
    return lists;
  }

  static RCRTCIWCustomLayout toLiveMixCustomLayout(@NonNull ReadableMap layout) {
    String id = layout.getString("id");
    int x = getInt(layout, "x", 0);
    int y = getInt(layout, "y", 0);
    int width = getInt(layout, "width", 480);
    int height = getInt(layout, "height", 640);
    RCRTCIWStreamType type;
    String tag = null;
    if (layout.hasKey("type"))
      type = RCRTCIWStreamType.values()[(getInt(layout, "type", 0))];
    else
      type = RCRTCIWStreamType.NORMAL;
    if (layout.hasKey("tag"))
      tag = layout.getString("tag");
    return new RCRTCIWCustomLayout(type, id, tag, x, y, width, height);
  }

  static WritableMap fromNetworkStats(@NonNull RCRTCIWNetworkStats stats) {
    WritableMap map = Arguments.createMap();
    map.putInt("type", stats.getType().ordinal());
    map.putString("ip", stats.getIp());
    map.putInt("sendBitrate", stats.getSendBitrate());
    map.putInt("receiveBitrate", stats.getReceiveBitrate());
    map.putInt("rtt", stats.getRtt());
    return map;
  }

  static WritableMap fromNetworkProbeStats(@NonNull RCRTCIWNetworkProbeStats stats) {
    WritableMap map = Arguments.createMap();
    map.putInt("qualityLevel", stats.getQualityLevel().ordinal());
    map.putInt("rtt", stats.getRtt());
    map.putDouble("packetLostRate", stats.getPacketLostRate());
    return map;
  }

  static WritableMap fromLocalAudioStats(@NonNull RCRTCIWLocalAudioStats stats) {
    WritableMap map = Arguments.createMap();
    map.putInt("codec", stats.getCodec().ordinal());
    map.putInt("bitrate", stats.getBitrate());
    map.putInt("volume", stats.getVolume());
    map.putDouble("packageLostRate", stats.getPackageLostRate());
    map.putInt("rtt", stats.getRtt());
    return map;
  }

  static WritableMap fromLocalVideoStats(@NonNull RCRTCIWLocalVideoStats stats) {
    WritableMap map = Arguments.createMap();
    map.putBoolean("tiny", stats.isTiny());
    map.putInt("codec", stats.getCodec().ordinal());
    map.putInt("bitrate", stats.getBitrate());
    map.putInt("fps", stats.getFps());
    map.putInt("width", stats.getWidth());
    map.putInt("height", stats.getHeight());
    map.putDouble("packageLostRate", stats.getPackageLostRate());
    map.putInt("rtt", stats.getRtt());
    return map;
  }

  static WritableMap fromRemoteAudioStats(@NonNull RCRTCIWRemoteAudioStats stats) {
    WritableMap map = Arguments.createMap();
    map.putInt("codec", stats.getCodec().ordinal());
    map.putInt("bitrate", stats.getBitrate());
    map.putInt("volume", stats.getVolume());
    map.putDouble("packageLostRate", stats.getPackageLostRate());
    map.putInt("rtt", stats.getRtt());
    return map;
  }

  static WritableMap fromRemoteVideoStats(@NonNull RCRTCIWRemoteVideoStats stats) {
    WritableMap map = Arguments.createMap();
    map.putInt("codec", stats.getCodec().ordinal());
    map.putInt("bitrate", stats.getBitrate());
    map.putInt("fps", stats.getFps());
    map.putInt("width", stats.getWidth());
    map.putInt("height", stats.getHeight());
    map.putDouble("packageLostRate", stats.getPackageLostRate());
    map.putInt("rtt", stats.getRtt());
    return map;
  }
}
